runtime/runtime_plugins.go

```diff
@@ -0,0 +1,216 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	loaded, err := loader.Filtered(params.Paths, params.Filter)
+	if err != nil {
+		return nil, err
+	}
+
+	store := inmem.New()
+
+	txn, err := store.NewTransaction(ctx, storage.WriteParams)
+	if err != nil {
+		return nil, err
+	}
+
+	// only register custom plugins if directory specified
+	if params.BuiltinDir != "" {
+		err = RegisterBuiltinsFromDir(params.BuiltinDir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := store.Write(ctx, txn, storage.AddOp, storage.Path{}, loaded.Documents); err != nil {
+		store.Abort(ctx, txn)
+		return nil, errors.Wrapf(err, "storage error")
+	}
+
+	if err := compileAndStoreInputs(ctx, store, txn, loaded.Modules, params.ErrorLimit); err != nil {
+		store.Abort(ctx, txn)
+		return nil, errors.Wrapf(err, "compile error")
+	}
+
+	if err := store.Commit(ctx, txn); err != nil {
+		return nil, errors.Wrapf(err, "storage error")
+	}
+
+	// register before init
+	if params.PluginDir != "" {
+		err = RegisterPluginsFromDir(params.PluginDir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	m, plugins, err := initPlugins(params.ID, store, params.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var decisionLogger func(context.Context, *server.Info)
+
+	if p, ok := plugins["decision_logs"]; ok {
+		decisionLogger = p.(*logs.Plugin).Log
+
+		if params.DecisionIDFactory == nil {
+			params.DecisionIDFactory = generateDecisionID
+		}
+	}
+
+	rt := &Runtime{
+		Store:          store,
+		Manager:        m,
+		Params:         params,
+		decisionLogger: decisionLogger,
+	}
+
+	return rt, nil
+}
+
+// RegisterBuiltinsFromDir recursively loads all custom builtins into OPA from dir. This function is idempotent.
+func RegisterBuiltinsFromDir(dir string) error {
+	return filepath.Walk(dir, walker("*.builtin.so", registerBuiltinFromFile))
+}
+
+// RegisterPluginsFromDir recursively loads all custom plugins into OPA from dir. This function is idempotent
+func RegisterPluginsFromDir(dir string) error {
+	return filepath.Walk(dir, walker("*.plugin.so", registerPluginFromFile))
+}
+
+// walker returns a walkfunc that performs handler on every file that matches the file glob pattern.
+// it skips all files that do not match pattern TODO: is this what we want?
+func walker(pattern string, handler func(string) error) filepath.WalkFunc {
+	walk := func(path string, f os.FileInfo, err error) error {
+		// if error occurs during traversal to path, exit and crash
+		if err != nil {
+			return err
+		}
+		// skip anything that is a directory
+		if f.IsDir() {
+			return nil
+		}
+
+		if ok, err := filepath.Match(pattern, filepath.Base(path)); err != nil {
+			return err
+		} else if !ok {
+			return nil
+		}
+
+		return handler(path)
+	}
+	return walk
+}
+
+// loads the builtin from a file path
+func registerBuiltinFromFile(path string) error {
+	mod, err := plugin.Open(path)
+	if err != nil {
+		return err
+	}
+	builtinSym, err := mod.Lookup("Builtin")
```

Comment:
Thinking about this a bit more...

The current design/implementation requires that users build a separate shared object for each built-in function/plugin they want to add. I could see this becoming a pain.

I think it would be better if either (a) the symbols we load can be slices or (b) we just call a single init function on the (Golang) plugin and it's up to the implementation to call ast.RegisterBuiltin, topdown.RegisterBuiltinFunc, etc. Option (b) is more flexible. Is there a compelling reason not to go that route?